src/scene/layer.js

```diff
@@ -400,15 +400,20 @@ Layer.prototype.addMeshInstances = function (meshInstances, skipShadowCasters) {
     var casters = this.shadowCasters;
     for (var i = 0; i < meshInstances.length; i++) {
         m = meshInstances[i];
+
         mat = m.material;
         if (mat.blendType === BLEND_NONE) {
             arr = this.opaqueMeshInstances;
         } else {
             arr = this.transparentMeshInstances;
         }
 
+        // test for meshInstance in both arrays, as material's alpha could have changed since LayerComposition's update to avoid duplicates
         // TODO - following uses of indexOf are expensive, to add 5000 meshInstances costs about 70ms on Mac. Consider using Set.
-        if (arr.indexOf(m) < 0) arr.push(m);
+        if (this.opaqueMeshInstances.indexOf(m) < 0 && this.transparentMeshInstances.indexOf(m) < 0) {
```

Comment:
perhaps you could optimise this in the general case: test the opaque array first, then semitrans array for opaque materials and visa-versa for semitrans materials?

Reply:
there is && in the middle, so both need to execute, not || where only one branch executes

Reply:
`&&` is also short-circuited though? if `a` is false in `a && b`, then `b` does not need to be evaluated.

Reply:
true, the optimization could be done. But I'm not sure it's worth the code complexity - we'll do proper Set based optimization instead I hope. Most meshes are opaque, so we can the benefit in most cases already.